From 65fc9c25122b6a679192e2fce73031d68a5cf8e0 Mon Sep 17 00:00:00 2001
From: Nikita Enin <n.enin@samsung.com>
Date: Sat, 29 Oct 2022 02:09:44 +0300
Subject: [PATCH 092/128] [caffe2, cmake] Axdimm shared libraries dependency
 for libtorch_cpu

Axdimm packages config to Dependencies in torch project
Add AxdimmAI, AxdimmSecure and AxdimmSgx dependencies to
libtorch_cpu

Resolves: AXDIMM-337

Signed-off-by: Nikita Enin <n.enin@samsung.com>
---
 caffe2/CMakeLists.txt    |  7 -------
 cmake/Dependencies.cmake | 25 +++++++++++++++++++++----
 2 files changed, 21 insertions(+), 11 deletions(-)

diff --git a/caffe2/CMakeLists.txt b/caffe2/CMakeLists.txt
index 474ad23c..309012d4 100644
--- a/caffe2/CMakeLists.txt
+++ b/caffe2/CMakeLists.txt
@@ -1374,18 +1374,11 @@ target_include_directories(torch_cpu SYSTEM PRIVATE "${Caffe2_DEPENDENCY_INCLUDE
 
 if (USE_AXDIMM)
   target_link_libraries(torch_axdimm PRIVATE ${Caffe2_AXDIMM_DEPENDENCY_LIBS})
-  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_ROOT_PATH}/build/ai/cmake")
-  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_ROOT_PATH}/build/secure/cmake")
-  find_package(AxdimmAI REQUIRED)
-  find_package(AxdimmSecure REQUIRED)
-  set(AXDIMM_LIBS Axdimm::AxdimmSecure Axdimm::AxdimmAI)
 
   if (USE_AXDIMM_SGX)
     target_compile_definitions(torch_axdimm PUBLIC -DSGX_ENABLE)
-    list(APPEND AXDIMM_LIBS Axdimm::AxdimmSecureSgx)
   endif ()
 
-  target_link_libraries(torch_axdimm PUBLIC ${AXDIMM_LIBS})
   target_include_directories(torch_axdimm PRIVATE ${Caffe2_AXDIMM_INCLUDE})
 endif ()
 
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index 47c61f2c..7fcf7572 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -679,10 +679,6 @@ if(USE_VULKAN)
   list(APPEND Caffe2_DEPENDENCY_LIBS ${Vulkan_LIBS})
 endif()
 
-if(USE_AXDIMM)
-  include_directories(SYSTEM ${CMAKE_AXDIMM_ROOT_PATH})
-endif()
-
 # ---[ gflags
 if(USE_GFLAGS)
   include(${CMAKE_CURRENT_LIST_DIR}/public/gflags.cmake)
@@ -1825,6 +1821,27 @@ set(TEMP_BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS})
 set(BUILD_SHARED_LIBS OFF CACHE BOOL "Build shared libs" FORCE)
 add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/fmt)
 
+# This defined after fmt processing to resolve fmt::fmt package
+# which is is required by axdimm libraries.
+# This package export is not used to link against fmt library by Axdimm libraries here
+# since Axdimm libraries are already built at this point and used solely
+# to define one, to stop cmake complaining.
+# Axdimm cmake files are required only to export axdimm library packages
+if(USE_AXDIMM)
+  include_directories(SYSTEM ${CMAKE_AXDIMM_ROOT_PATH})
+  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_ROOT_PATH}/build/ai/cmake")
+  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_ROOT_PATH}/build/secure/cmake")
+  find_package(AxdimmAI REQUIRED)
+  find_package(AxdimmSecure REQUIRED)
+  set(AXDIMM_LIBS Axdimm::AxdimmSecure Axdimm::AxdimmAI)
+
+  if (USE_AXDIMM_SGX)
+    list(APPEND AXDIMM_LIBS Axdimm::AxdimmSecureSgx)
+  endif ()
+  list(APPEND Caffe2_DEPENDENCY_LIBS ${AXDIMM_LIBS})
+  list(APPEND Caffe2_AXDIMM_DEPENDENCY_LIBS ${AXDIMM_LIBS})
+endif()
+
 # Disable compiler feature checks for `fmt`.
 #
 # CMake compiles a little program to check compiler features. Some of our build
-- 
2.34.1

